blazar/plugins/devices/device_plugin.py

```diff
@@ -710,10 +736,210 @@ def _allocations_to_remove(self, dates_before, dates_after, max_devices,
+                continue
+
+            for allocation in [alloc for alloc
+                               in reservation['device_allocations']
+                               if alloc['device_id'] in device_ids]:
+                if self._reallocate(allocation):
+                    if reservation['status'] == status.reservation.ACTIVE:
+                        if reservation['id'] not in reservation_flags:
+                            reservation_flags[reservation['id']] = {}
+                        reservation_flags[reservation['id']].update(
+                            {'resources_changed': True})
+                else:
+                    if reservation['id'] not in reservation_flags:
+                        reservation_flags[reservation['id']] = {}
+                    reservation_flags[reservation['id']].update(
+                        {'missing_resources': True})
+
+        return reservation_flags
+
+
+class DeviceMonitorPlugin(base.BaseMonitorPlugin):
+    """Monitor plugin for device resource."""
+
+    # Singleton design pattern
+    _instance = None
+
+    def __new__(cls):
+        if not cls._instance:
+            cls._instance = super(DeviceMonitorPlugin, cls).__new__(cls)
+            cls._instance.healing_handlers = []
+            cls._instance.plugins = _get_plugins()
+            super(DeviceMonitorPlugin, cls._instance).__init__()
+        return cls._instance
+
+    def __init__(self):
+        """Do nothing.
+
+        This class uses the Singleton design pattern and an instance of this
+        class is generated and initialized in __new__().
+        """
+        pass
+
+    def register_healing_handler(self, handler):
+        self.healing_handlers.append(handler)
+
+    def is_notification_enabled(self):
+        """Check if the notification monitor is enabled."""
+        return CONF[plugin.RESOURCE_TYPE].enable_notification_monitor
+
+    def get_notification_event_types(self):
+        """Get event types of notification messages to handle."""
+        return []
+
+    def get_notification_topics(self):
+        """Get topics of notification to subscribe to."""
+        return CONF[plugin.RESOURCE_TYPE].notification_topics
+
+    def notification_callback(self, event_type, payload):
+        """Handle a notification message.
+        It is used as a callback of a notification-based resource monitor.
+        :param event_type: an event type of a notification.
+        :param payload: a payload of a notification.
+        :return: a dictionary of {reservation id: flags to update}
+                 e.g. {'de27786d-bd96-46bb-8363-19c13b2c6657':
+                       {'missing_resources': True}}
+        """
+        LOG.trace('Handling a notification...')
+        reservation_flags = {}
+
+        for device_driver in self.plugins.keys():
+            try:
+                reservation_flags.update(
+                    self.plugins[device_driver].notification_callback(
```

Comment:
We don't use the notification-style monitoring, so this won't blow up, but it also won't work. What about removing this and the branching logic? I'm not even sure what events Zun will emit that we could use here.